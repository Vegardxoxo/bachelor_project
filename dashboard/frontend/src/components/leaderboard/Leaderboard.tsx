import * as React from 'react';
import {useEffect} from 'react';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import {Box, Grid, Stack} from '@mui/material';
import ActiveLastBreadcrumb from '../dashboard/ActivateLastBreadcrumb/ActivateLastBreadcrumb';
import {useRecoilValue} from 'recoil';
import {userAtom} from '../../globalVariables/variables';

interface Leaderboard {
    organization: string;
    active_percentage: number;
    rank: number;
}

/**
 * Renders the Leaderboard component.
 * The Leaderboard component shows the user's unit rank and the top 25 units based on their active percentage.
 */
export function Leaderboard() {
    // State to store fetched leaderboard data
    const [data, setData] = React.useState<Leaderboard[]>([]);
    // Retrieve access token and user information
    const accessToken = localStorage.getItem('access');
    const userInfo = useRecoilValue(userAtom);

    // Fetch leaderboard data from the API
    useEffect(() => {
        const fetchData = async () => {
            try {
                const response = await fetch(
                    `http://127.0.0.1:8000/api/licenses/leaderboard/?organization=${userInfo.organization}`,
                    {
                        method: 'GET',
                        headers: {
                            'content-type': 'application/json',
                            authorization: `Bearer ${accessToken}`
                        }
                    }
                );
                const data = await response.json();
                if (response.ok) {
                    setData(data.leaderboard);
                }
            } catch (error) {
                console.error('Error while fetching data:', error);
            }
        };
        fetchData();
    }, []);


    /*Renders the Leaderboard table. The first table shows the users unit rank.
    And the second table shows the top 25 units based on their active percentage */
    return (
        <>
            <div>
                <Grid>
                    <ActiveLastBreadcrumb/>
                </Grid>
                <Box
                    data-testid="leaderboard"
                    style={{
                        display: 'flex',
                        justifyContent: 'center',
                        alignContent: 'center',
                        marginTop: '20px',
                        marginLeft: '1.8%',
                        width: '96.3%'
                    }}>
                    <Grid justifyContent={'center'} alignItems={'center'} width={'75%'}>
                        <Stack direction={'column'} width={'90%'}>
                            {/* Leaderboard title and description */}
                            <h2
                                style={{
                                    fontFamily: 'Source Sans Pro, sans serif',
                                    fontSize: '30pt',
                                    marginTop: -0.6,
                                    fontWeight: 400
                                }}>
                                Ledertavle
                            </h2>
                            <p
                                style={{
                                    fontStyle: 'italic',
                                    marginTop: '-10px',
                                    marginBottom: '40px'
                                }}>
                                Ledertavlen gir deg en visuell representasjon av hvordan din
                                enhet presterer i forhold til andre enheter, basert på andelen
                                av aktive lisenser. Topplisten gir en oversikt over de 25
                                enhetene med flest aktive lisenser, og indikerer dermed hvilke
                                enheter som har færrest ubrukte lisenser.{' '}
                            </p>
                        </Stack>

                        {/* User's unit rank table */}
                        <p>Din plassering</p>
                        <TableContainer component={Paper}>
                            <Table
                                sx={{
                                    minWidth: 650,
                                    fontFamily: 'Source Sans Pro, sans serif',
                                    fontWeight: '700',
                                    backgroundColor: '#C5DAF2'
                                }}
                                aria-label="simple table">
                                <colgroup>
                                    <col style={{width: '10%'}}/>
                                    <col style={{width: '80%'}}/>
                                    <col style={{width: '10%'}}/>
                                </colgroup>
                                <TableHead>
                                    <TableRow
                                        sx={{fontWeight: 'bold', padding: '10px'}}></TableRow>
                                </TableHead>
                                <TableBody>
                                    {/* Render user's unit rank */}
                                    {data.map((row, index) => {
                                        if (userInfo.organization === row.organization) {
                                            return (
                                                <TableRow
                                                    key={index}
                                                    sx={{
                                                        '&:last-child td, &:last-child th': {
                                                            border: 0,
                                                            backgroundColor: '#80cc9f'
                                                        }
                                                    }}>
                                                    <TableCell component="th" scope="row">
                                                        {row.rank}
                                                    </TableCell>
                                                    <TableCell align="left">{row.organization}</TableCell>
                                                    <TableCell align="center">
                                                        {row.active_percentage}
                                                    </TableCell>
                                                </TableRow>
                                            );
                                        } else {
                                            return null;
                                        }
                                    })}
                                </TableBody>
                            </Table>
                        </TableContainer>

                        <div style={{padding: '10px'}}></div>
                        {/* Top 25 units table */}
                        <p>Topplisten</p>
                        <TableContainer component={Paper}>
                            <Table
                                sx={{
                                    minWidth: 650,
                                    fontFamily: 'Source Sans Pro, sans serif',
                                    borderRadius: '50px',
                                    fontWeight: '700'
                                }}
                                aria-label="simple table">
                                <colgroup>
                                    <col style={{width: '10%'}}/>
                                    <col style={{width: '80%'}}/>
                                    <col style={{width: '10%'}}/>
                                </colgroup>
                                <TableHead>
                                    <TableRow>
                                        <TableCell></TableCell>
                                        <TableCell
                                            align="left"
                                            sx={{
                                                fontFamily: 'Source Sans Pro, sans serif',
                                                fontSize: 'medium',
                                                fontWeight: 900
                                            }}>
                                            Enhet
                                        </TableCell>
                                        <TableCell
                                            align="center"
                                            sx={{
                                                fontFamily: 'Source Sans Pro, sans serif',
                                                fontSize: 'medium',
                                                fontWeight: 900
                                            }}>
                                            Prosent
                                        </TableCell>
                                    </TableRow>
                                </TableHead>
                                <TableBody>
                                    {/* Render top 25 units */}
                                    {data.slice(0, 25).map((row, index) => (
                                        <TableRow
                                            key={index}
                                            sx={{
                                                '&:last-child td, &:last-child th': {border: 0}
                                            }}>
                                            <TableCell
                                                component="th"
                                                scope="row"
                                                sx={{fontWeight: 500}}>
                                                {row.rank}
                                            </TableCell>
                                            <TableCell align="left">{row.organization}</TableCell>
                                            <TableCell align="center">
                                                {row.active_percentage}
                                            </TableCell>
                                        </TableRow>
                                    ))}
                                </TableBody>
                            </Table>
                        </TableContainer>
                    </Grid>
                </Box>
            </div>
        </>
    );
}
