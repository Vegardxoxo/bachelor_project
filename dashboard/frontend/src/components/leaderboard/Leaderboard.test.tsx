import {render, screen, waitFor} from '@testing-library/react';
import {Leaderboard} from './Leaderboard';
import {MemoryRouter} from 'react-router-dom';
import {RecoilRoot} from 'recoil';
import '@testing-library/jest-dom/extend-expect';
import renderer from "react-test-renderer";
import React from "react";

jest.mock('node-fetch');

describe('Leaderboard component', () => {
    beforeEach(async () => {
        const mockData = {
            leaderboard: [
                {
                    organization: 'IT-tjenester',
                    active_percentage: 50.1,
                    rank: 5
                },
                {
                    organization: 'Berg skole',
                    active_percentage: 7.6,
                    rank: 101
                }
            ]
        };
        global.fetch = jest.fn().mockImplementation(() =>
            Promise.resolve({
                ok: true,
                json: () => Promise.resolve(mockData)
            })
        );

        render(
            <MemoryRouter>
                <RecoilRoot>
                    <Leaderboard/>
                </RecoilRoot>
            </MemoryRouter>
        );

        // Wait for the leaderboard data to be fetched and rendered
        await waitFor(() => screen.getByText('IT-tjenester'));
    });

    test('matches snapshot', async () => {
        const testRenderer = renderer.create(
            <MemoryRouter>
                <RecoilRoot>
                    <Leaderboard/>
                </RecoilRoot>
            </MemoryRouter>
        );
        expect(testRenderer.toJSON()).toMatchSnapshot();
    })
    test('displays the correct organization', () => {
        const org1 = screen.getByText('IT-tjenester');
        const org2 = screen.getByText('Berg skole');
        expect(org1).toBeInTheDocument();
        expect(org2).toBeInTheDocument();
    });

    test('displays the correct percentage', () => {
        const org1Perc = screen.getByText('50.1');
        expect(org1Perc).toBeInTheDocument();
        const org2Perc = screen.getByText('7.6');
        expect(org2Perc).toBeInTheDocument();
    });

    test('displays the correct ranking', () => {
        const org1Rank = screen.getByText('5');
        expect(org1Rank).toBeInTheDocument();
        const org2Rank = screen.getByText('101');
        expect(org2Rank).toBeInTheDocument();
    });

    test('displays the correct header', () => {
        expect(screen.getByText('Topplisten')).toBeInTheDocument();
    });
});
