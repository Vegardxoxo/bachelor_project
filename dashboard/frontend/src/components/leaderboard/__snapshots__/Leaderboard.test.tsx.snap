// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard component matches snapshot 1`] = `
<div>
  <div
    className="MuiGrid-root css-vj1n65-MuiGrid-root"
  >
    <nav
      aria-label="breadcrumb"
      className="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      id="breadcrumbs"
    >
      <ol
        className="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
      >
        <li
          className="MuiBreadcrumbs-li"
        >
          <a
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-le90w3-MuiTypography-root-MuiLink-root"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
          >
            Dashboard
          </a>
        </li>
      </ol>
    </nav>
  </div>
  <div
    className="MuiBox-root css-0"
    data-testid="leaderboard"
    style={
      Object {
        "alignContent": "center",
        "display": "flex",
        "justifyContent": "center",
        "marginLeft": "1.8%",
        "marginTop": "20px",
        "width": "96.3%",
      }
    }
  >
    <div
      className="MuiGrid-root css-1bpij53-MuiGrid-root"
    >
      <div
        className="css-70ur0o-MuiStack-root"
      >
        <h2
          style={
            Object {
              "fontFamily": "Source Sans Pro, sans serif",
              "fontSize": "30pt",
              "fontWeight": 400,
              "marginTop": -0.6,
            }
          }
        >
          Ledertavle
        </h2>
        <p
          style={
            Object {
              "fontStyle": "italic",
              "marginBottom": "40px",
              "marginTop": "-10px",
            }
          }
        >
          Ledertavlen gir deg en visuell representasjon av hvordan din enhet presterer i forhold til andre enheter, basert på andelen av aktive lisenser. Topplisten gir en oversikt over de 25 enhetene med flest aktive lisenser, og indikerer dermed hvilke enheter som har færrest ubrukte lisenser.
           
        </p>
      </div>
      <p>
        Din plassering
      </p>
      <div
        className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
      >
        <table
          aria-label="simple table"
          className="MuiTable-root css-fm8wf7-MuiTable-root"
          role={null}
        >
          <colgroup>
            <col
              style={
                Object {
                  "width": "10%",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "10%",
                }
              }
            />
          </colgroup>
          <thead
            className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1yxvnga-MuiTableRow-root"
              role={null}
            />
          </thead>
          <tbody
            className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            role={null}
          />
        </table>
      </div>
      <div
        style={
          Object {
            "padding": "10px",
          }
        }
      />
      <p>
        Topplisten
      </p>
      <div
        className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
      >
        <table
          aria-label="simple table"
          className="MuiTable-root css-i6j12u-MuiTable-root"
          role={null}
        >
          <colgroup>
            <col
              style={
                Object {
                  "width": "10%",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <col
              style={
                Object {
                  "width": "10%",
                }
              }
            />
          </colgroup>
          <thead
            className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              />
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-gvdt8c-MuiTableCell-root"
                scope="col"
              >
                Enhet
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1hzidc9-MuiTableCell-root"
                scope="col"
              >
                Prosent
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            role={null}
          />
        </table>
      </div>
    </div>
  </div>
</div>
`;
